builder/amazon/ebs/step_create_ami.go

```diff
@@ -48,7 +50,22 @@ func (s *stepCreateAMI) Run(ctx context.Context, state multistep.StateBag) multi
 		BlockDeviceMappings: config.AMIMappings.BuildEC2BlockDeviceMappings(),
 	}
 
-	createResp, err := ec2conn.CreateImage(createOpts)
+	var createResp *ec2.CreateImageOutput
+	var err error
+
+	err = retry.Config{
+		Tries: 11,
```

Comment:
Tried the test build and the error is slightly different:

```
ami: Error creating AMI: retry count exhausted. Last err: InvalidParameterValue: Invalid value 'i-00a737b5f100ce79d' for instanceId. Instance is not in state 'running' or 'stopped'\n==> ami: 	status code: 400%!(PACKER_COMMA) request id: 2144ba32-b620-4042-bf19-e7e6d522ee80
1594849969,,ui,say,==> ami: Provisioning step had errors: Running the cleanup provisioner%!(PACKER_COMMA) if present...
```

I tried setting `AWS_MAX_ATTEMPTS = 100` and `AWS_POLL_DELAY_SECONDS = 30` but still getting the same error.

I wonder if this line will make tries number to 11 no matter what you set to those 2 env vars?

Reply:
yeah the retry here is not related to the env vars, which get used when declaring waiters. 

I'm surprised though. If this were eventual consistency related, I'd have expected this wait loop to be enough. Can you set the environment variable PACKER_LOG=1 to turn on versbose logging and share full logs on the original issue? I'd also be interested to know how long Packer is spending around these lines: 

```
==> amazon-ebs: Stopping the source instance...
amazon-ebs: Stopping instance
==> amazon-ebs: Waiting for the instance to stop...
==> amazon-ebs: Creating AMI test from instance i-0dfeb326a14cb2aa4```

Reply:
2020/07/16 00:03:56 1594857836,,ui,say,==> ami: Stopping the source instance...
1594857836,,ui,say,==> ami: Stopping the source instance...
1594857836,,ui,message,    ami: Stopping instance
2020/07/16 00:03:56 1594857836,,ui,message,    ami: Stopping instance
2020/07/16 00:03:56 1594857836,,ui,say,==> ami: Waiting for the instance to stop...
1594857836,,ui,say,==> ami: Waiting for the instance to stop...
1594858168,,ui,say,==> ami: Creating AMI xxx from instance i-xxx
2020/07/16 00:09:28 1594858168,,ui,say,==> ami: Creating AMI xxx from instance i-xxx
2020/07/16 00:09:28 packer_test plugin: Retryable error: InvalidParameterValue: Invalid value 'i-xxx' for instanceId. Instance is not in state 'running' or 'stopped'
2020/07/16 00:09:28 packer_test plugin: 	status code: 400, request id: x
2020/07/16 00:09:29 packer_test plugin: Retryable error: InvalidParameterValue: Invalid value 'i-xxx' for instanceId. Instance is not in state 'running' or 'stopped'
2020/07/16 00:09:29 packer_test plugin: 	status code: 400, request id: x
2020/07/16 00:09:29 packer_test plugin: Retryable error: InvalidParameterValue: Invalid value 'i-xxx' for instanceId. Instance is not in state 'running' or 'stopped'
2020/07/16 00:09:29 packer_test plugin: 	status code: 400, request id: x
2020/07/16 00:09:30 packer_test plugin: Retryable error: InvalidParameterValue: Invalid value 'i-xxx' for instanceId. Instance is not in state 'running' or 'stopped'
2020/07/16 00:09:30 packer_test plugin: 	status code: 400, request id: x
2020/07/16 00:09:32 packer_test plugin: Retryable error: InvalidParameterValue: Invalid value 'i-xxx' for instanceId. Instance is not in state 'running' or 'stopped'
2020/07/16 00:09:32 packer_test plugin: 	status code: 400, request id: x
2020/07/16 00:09:35 packer_test plugin: Retryable error: InvalidParameterValue: Invalid value 'i-xxx' for instanceId. Instance is not in state 'running' or 'stopped'
2020/07/16 00:09:35 packer_test plugin: 	status code: 400, request id: x
2020/07/16 00:09:42 packer_test plugin: Retryable error: InvalidParameterValue: Invalid value 'i-xxx' for instanceId. Instance is not in state 'running' or 'stopped'
2020/07/16 00:09:42 packer_test plugin: 	status code: 400, request id: x
2020/07/16 00:09:55 packer_test plugin: Retryable error: InvalidParameterValue: Invalid value 'i-xxx' for instanceId. Instance is not in state 'running' or 'stopped'
2020/07/16 00:09:55 packer_test plugin: 	status code: 400, request id: x
2020/07/16 00:10:21 packer_test plugin: Retryable error: InvalidParameterValue: Invalid value 'i-xxx' for instanceId. Instance is not in state 'running' or 'stopped'
2020/07/16 00:10:21 packer_test plugin: 	status code: 400, request id: x
2020/07/16 00:10:51 packer_test plugin: Retryable error: InvalidParameterValue: Invalid value 'i-xxx' for instanceId. Instance is not in state 'running' or 'stopped'
2020/07/16 00:10:51 packer_test plugin: 	status code: 400, request id: x
2020/07/16 00:11:21 packer_test plugin: Retryable error: InvalidParameterValue: Invalid value 'i-xxx' for instanceId. Instance is not in state 'running' or 'stopped'
2020/07/16 00:11:21 packer_test plugin: 	status code: 400, request id: x
2020/07/16 00:11:51 1594858311,,ui,error,==> ami: Error creating AMI: retry count exhausted. Last err: InvalidParameterValue: Invalid value 'i-xxx' for instanceId. Instance is not in state 'running' or 'stopped'\n==> ami: 	status code: 400%!(PACKER_COMMA) request id: x
1594858311,,ui,error,==> ami: Error creating AMI: retry count exhausted. Last err: InvalidParameterValue: Invalid value 'i-xxx' for instanceId. Instance is not in state 'running' or 'stopped'\n==> ami: 	status code: 400%!(PACKER_COMMA) request id: x
1594858311,,ui,say,==> ami: Provisioning step had errors: Running the cleanup provisioner%!(PACKER_COMMA) if present...

Reply:
This is so weird.

Reply:
Those EC2 metal boxes do take long time to power on/off. IMO we probably need longer waiting time and/or more retries

Reply:
I believe we need longer waiting time / retries for this

I launched the packer build with the `-debug` flag so I could single step it. I waited 15 min after `StepStopEBSBackedInstance` and the build works

```
    ami: PLAY RECAP *********************************************************************
    ami: default                    : ok=12   changed=11   unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
    ami:
2020/07/20 20:32:49 packer-provisioner-ansible plugin: shutting down the SSH proxy
2020/07/20 20:32:49 [INFO] (telemetry) ending ansible
==> ami: Pausing after run of step 'StepProvision'. Press enter to continue. 
==> ami: Trying to remove ephemeral keys from authorized_keys files
2020/07/20 20:36:43 packer-builder-amazon-ebs plugin: [DEBUG] Opening new ssh session
2020/07/20 20:36:43 packer-builder-amazon-ebs plugin: [DEBUG] starting remote command: sed -i.bak '/ packer_5f15e763-7cae-3838-65a1-8df87abe8a67$/d' ~/.ssh/authorized_keys; rm ~/.ssh/authorized_keys.bak
2020/07/20 20:36:43 packer-builder-amazon-ebs plugin: [DEBUG] Opening new ssh session
2020/07/20 20:36:43 packer-builder-amazon-ebs plugin: [DEBUG] starting remote command: sudo sed -i.bak '/ packer_5f15e763-7cae-3838-65a1-8df87abe8a67$/d' /root/.ssh/authorized_keys; sudo rm /root/.ssh/authorized_keys.bak
==> ami: Pausing after run of step 'StepCleanupTempKeys'. Press enter to continue. 
==> ami: Stopping the source instance...
    ami: Stopping instance
==> ami: Waiting for the instance to stop...
==> ami: Pausing after run of step 'StepStopEBSBackedInstance'. Press enter to continue. 

(I waited 15min here)

==> ami: Pausing after run of step 'StepModifyEBSBackedInstance'. Press enter to continue. 
==> ami: Pausing after run of step 'StepDeregisterAMI'. Press enter to continue. 
==> ami: Creating AMI xxx from instance i-xxx
    ami: AMI: ami-xxx
==> ami: Waiting for AMI to become ready...
==> ami: Pausing after run of step 'stepCreateAMI'. Press enter to continue. 
==> ami: Pausing after run of step 'StepAMIRegionCopy'. Press enter to continue. 
==> ami: Pausing after run of step 'StepModifyAMIAttributes'. Press enter to continue. 
==> ami: Adding tags to AMI (ami-xxx)...
==> ami: Tagging snapshot: snap-xxx
```

Reply:
Ah, nice, setting tries to `0` will not limit on the number of retires but will leave when ctx is cancelled. May be creating a context with a timeout of 15min will solve this:
```suggestion
		Tries: 0,
```

Reply:
The thing that's confusing me isn't that it takes a long time to stop -- it's that the waiter being called in the part of the code that says "Waiting for the instance to stop..." should be the one that waits a long time and has to do a bunch of retries, not this "creating image" step. We need to figure out why that waiter is passing when the instance clearly isn't stopped.

Reply:
IIRC When I was doing single step, I saw `==> ami: Pausing after run of step 'StepStopEBSBackedInstance'. Press enter to continue. ` right after AWS console showing the instance is stopped (I am refreshing page every few sec). So IMO the waiter did nothing wrong, it is just some special behaviour of AWS's metal instances

Reply:
So maybe for some reason the eventual consistency is more... "eventual" with the metal instances. Okay. I'll tweak this code again to make the retries last longer.

Reply:
```
2020/07/22 00:55:31 1595379331,,ui,say,==> ami: Waiting for the instance to stop...
1595379331,,ui,say,==> ami: Waiting for the instance to stop...
1595379664,,ui,say,==> ami: Creating AMI xxx from instance xxx
2020/07/22 01:01:04 1595379664,,ui,say,==> ami: Creating AMI xxx from instance xxx
1595379664,,ui,message,    ami: AMI: ami-xxx
2020/07/22 01:01:04 1595379664,,ui,message,    ami: AMI: ami-xxx
2020/07/22 01:01:04 1595379664,,ui,say,==> ami: Waiting for AMI to become ready...
```
It works!
Thanks @SwampDragons @azr !